group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -874,4 +1045,1265 @@ public void replay(
+                group.allMembers().forEach(member -> {
+
+                    List<JoinGroupResponseData.JoinGroupResponseMember> members = Collections.emptyList();
+                    if (group.isLeader(member.memberId())) {
+                        members = group.currentGenericGroupMembers();
+                    }
+
+                    JoinGroupResponseData response = new JoinGroupResponseData()
+                        .setMembers(members)
+                        .setMemberId(member.memberId())
+                        .setGenerationId(group.generationId())
+                        .setProtocolName(group.protocolName().orElse(null))
+                        .setProtocolType(group.protocolType().orElse(null))
+                        .setLeader(group.leaderOrNull())
+                        .setSkipAssignment(false)
+                        .setErrorCode(Errors.NONE.code());
+
+                    group.completeJoinFuture(member, response);
+                    rescheduleGenericGroupMemberHeartbeat(group, member);
+                    member.setIsNew(false);
+
+                    group.addPendingSyncMember(member.memberId());
+                });
+
+                schedulePendingSync(group);
+            }
+        }
+
+        return EMPTY_RESULT;
+    }
+
+    /**
+     * Wait for sync requests for the group.
+     *
+     * @param group The group.
+     */
+    private void schedulePendingSync(GenericGroup group) {
+        timer.schedule(syncKey(group.groupId()),
+            group.rebalanceTimeoutMs(),
+            TimeUnit.MILLISECONDS,
+            () -> expirePendingSync(group, group.generationId()));
+    }
+
+    /**
+     * Invoked when the heartbeat operation is expired from the timer. Possibly remove the member and
+     * try complete the join phase.
+     *
+     * @param group     The group.
+     * @param memberId  The member id.
+     *
+     * @return The coordinator result that will be appended to the log.
+     */
+    private CoordinatorResult<CompletableFuture<Errors>, Record> expireGenericGroupMemberHeartbeat(
+        GenericGroup group,
+        String memberId
+    ) {
+        if (group.isInState(DEAD)) {
+            log.info("Received notification of heartbeat expiration for member {} after group {} " +
+                    "had already been unloaded or deleted.",
+                memberId, group.groupId());
+
+        } else if (group.isPendingMember(memberId)) {
+            log.info("Pending member {} in group {} has been removed after session timeout expiration.",
+                memberId, group.groupId());
+
+            return removePendingMemberAndUpdateGenericGroup(group, memberId);
+        } else if (!group.hasMemberId(memberId)) {
+            log.debug("Member {} has already been removed from the group.", memberId);
+        } else {
+            GenericGroupMember member = group.member(memberId);
+            if (!member.hasSatisfiedHeartbeat()) {
```

Comment:
if the member is awaiting on a join/sync response then we can't remove the member on hb expiration right?

Reply:
i think we can cancel the existing heartbeat when a member rejoins during a rebalance. this will still expire if the member does not rejoin which is what we want. also, moving the new member join timeout to a different key will help remove this `hasSatisfiedHeartbeat`.

is that what you had in mind?